File: scripts/fix-page-placement.mjs
/**
 * Fix script: moves <Page> from inside .map() callbacks to the top-level return.
 *
 * The migrate-misc-bulk.mjs script sometimes inserts <Page> inside a .map()
 * callback's return() instead of the component's top-level return(). This
 * script detects and fixes that pattern.
 *
 * Usage:
 *   node scripts/fix-page-placement.mjs <file-path>
 *   node scripts/fix-page-placement.mjs --all   (fixes all recently migrated files)
 */

import { readFileSync, writeFileSync, readdirSync } from "fs";
import { join, resolve } from "path";

const ROOT = resolve(import.meta.dirname, "..");

function findPages(dir) {
  const results = [];
  try {
    for (const entry of readdirSync(dir, { withFileTypes: true })) {
      const full = join(dir, entry.name);
      if (entry.isDirectory()) results.push(...findPages(full));
      else if (entry.name === "page.tsx") results.push(full);
    }
  } catch {}
  return results;
}

/**
 * Check if a file has the "Page inside map" problem.
 * Heuristic: the file has <Page> but the <Page> tag appears AFTER a .map(
 * or .map((  line without a closing </Page> before the next .map(
 */
function hasPageInMapProblem(content) {
  if (!content.includes("@/components/ui/Page")) return false;
  
  // Find the position of <Page in the file
  const pageIdx = content.indexOf("<Page");
  if (pageIdx === -1) return false;
  
  // Check if there's a .map( before the <Page that doesn't have a closing paren
  // before the <Page
  const beforePage = content.substring(0, pageIdx);
  
  // Look for the last 'return (' before <Page
  const lastReturnBeforePage = beforePage.lastIndexOf("return (");
  if (lastReturnBeforePage === -1) return false;
  
  // Check if this return is inside a .map() callback
  // by looking for .map( before this return
  const beforeReturn = beforePage.substring(0, lastReturnBeforePage);
  const lastMapBeforeReturn = Math.max(
    beforeReturn.lastIndexOf(".map(("),
    beforeReturn.lastIndexOf(".map(d =>"),
    beforeReturn.lastIndexOf(".map((d,"),
    beforeReturn.lastIndexOf(".map((kpi"),
    beforeReturn.lastIndexOf(".map((item"),
    beforeReturn.lastIndexOf(".map((row"),
    beforeReturn.lastIndexOf(".map((plan"),
    beforeReturn.lastIndexOf(".map((step"),
    beforeReturn.lastIndexOf(".map((q,"),
    beforeReturn.lastIndexOf(".map((opt"),
    beforeReturn.lastIndexOf(".map((tab"),
    beforeReturn.lastIndexOf(".map((s,"),
    beforeReturn.lastIndexOf(".map((t,"),
    beforeReturn.lastIndexOf(".map((c,"),
    beforeReturn.lastIndexOf(".map((e,"),
    beforeReturn.lastIndexOf(".map((n,"),
    beforeReturn.lastIndexOf(".map((f,"),
    beforeReturn.lastIndexOf(".map((m,"),
    beforeReturn.lastIndexOf(".map((p,"),
    beforeReturn.lastIndexOf(".map((r,"),
    beforeReturn.lastIndexOf(".map((v,"),
    beforeReturn.lastIndexOf(".map((w,"),
    beforeReturn.lastIndexOf(".map((x,"),
    beforeReturn.lastIndexOf(".map((y,"),
    beforeReturn.lastIndexOf(".map((z,"),
    beforeReturn.lastIndexOf(".map((i,"),
    beforeReturn.lastIndexOf(".map((j,"),
    beforeReturn.lastIndexOf(".map((k,"),
    beforeReturn.lastIndexOf(".map((l,"),
    beforeReturn.lastIndexOf(".map((a,"),
    beforeReturn.lastIndexOf(".map((b,"),
    beforeReturn.lastIndexOf(".map((g,"),
    beforeReturn.lastIndexOf(".map((h,"),
    beforeReturn.lastIndexOf(".map((u,"),
    beforeReturn.lastIndexOf(".map((val"),
    beforeReturn.lastIndexOf(".map((dept"),
    beforeReturn.lastIndexOf(".map((emp"),
    beforeReturn.lastIndexOf(".map((user"),
    beforeReturn.lastIndexOf(".map((ticket"),
    beforeReturn.lastIndexOf(".map((award"),
    beforeReturn.lastIndexOf(".map((survey"),
    beforeReturn.lastIndexOf(".map((event"),
    beforeReturn.lastIndexOf(".map((benefit"),
    beforeReturn.lastIndexOf(".map((card"),
    beforeReturn.lastIndexOf(".map((link"),
    beforeReturn.lastIndexOf(".map((metric"),
    beforeReturn.lastIndexOf(".map((stat"),
    beforeReturn.lastIndexOf(".map((doc"),
    beforeReturn.lastIndexOf(".map((file"),
    beforeReturn.lastIndexOf(".map((role"),
    beforeReturn.lastIndexOf(".map((team"),
    beforeReturn.lastIndexOf(".map((goal"),
    beforeReturn.lastIndexOf(".map((task"),
    beforeReturn.lastIndexOf(".map((action"),
    beforeReturn.lastIndexOf(".map((phase"),
    beforeReturn.lastIndexOf(".map((stage"),
    beforeReturn.lastIndexOf(".map((level"),
    beforeReturn.lastIndexOf(".map((type"),
    beforeReturn.lastIndexOf(".map((cat"),
    beforeReturn.lastIndexOf(".map((tag"),
    beforeReturn.lastIndexOf(".map((badge"),
    beforeReturn.lastIndexOf(".map((chip"),
    beforeReturn.lastIndexOf(".map((filter"),
    beforeReturn.lastIndexOf(".map((col"),
    beforeReturn.lastIndexOf(".map((row"),
    beforeReturn.lastIndexOf(".map((cell"),
    beforeReturn.lastIndexOf(".map((node"),
    beforeReturn.lastIndexOf(".map((edge"),
    beforeReturn.lastIndexOf(".map((point"),
    beforeReturn.lastIndexOf(".map((bar"),
    beforeReturn.lastIndexOf(".map((line"),
    beforeReturn.lastIndexOf(".map((area"),
    beforeReturn.lastIndexOf(".map((pie"),
    beforeReturn.lastIndexOf(".map((slice"),
    beforeReturn.lastIndexOf(".map((segment"),
    beforeReturn.lastIndexOf(".map((section"),
    beforeReturn.lastIndexOf(".map((group"),
    beforeReturn.lastIndexOf(".map((cluster"),
    beforeReturn.lastIndexOf(".map((bucket"),
    beforeReturn.lastIndexOf(".map((bin"),
    beforeReturn.lastIndexOf(".map((range"),
    beforeReturn.lastIndexOf(".map((period"),
    beforeReturn.lastIndexOf(".map((month"),
    beforeReturn.lastIndexOf(".map((week"),
    beforeReturn.lastIndexOf(".map((day"),
    beforeReturn.lastIndexOf(".map((hour"),
    beforeReturn.lastIndexOf(".map((minute"),
    beforeReturn.lastIndexOf(".map((second"),
    beforeReturn.lastIndexOf(".map((year"),
    beforeReturn.lastIndexOf(".map((quarter"),
    beforeReturn.lastIndexOf(".map((semester"),
    beforeReturn.lastIndexOf(".map((season"),
    beforeReturn.lastIndexOf(".map((cycle"),
    beforeReturn.lastIndexOf(".map((sprint"),
    beforeReturn.lastIndexOf(".map((milestone"),
    beforeReturn.lastIndexOf(".map((checkpoint"),
    beforeReturn.lastIndexOf(".map((waypoint"),
    beforeReturn.lastIndexOf(".map((stop"),
    beforeReturn.lastIndexOf(".map((station"),
    beforeReturn.lastIndexOf(".map((location"),
    beforeReturn.lastIndexOf(".map((place"),
    beforeReturn.lastIndexOf(".map((site"),
    beforeReturn.lastIndexOf(".map((venue"),
    beforeReturn.lastIndexOf(".map((address"),
    beforeReturn.lastIndexOf(".map((city"),
    beforeReturn.lastIndexOf(".map((country"),
    beforeReturn.lastIndexOf(".map((region"),
    beforeReturn.lastIndexOf(".map((zone"),
    beforeReturn.lastIndexOf(".map((area"),
    beforeReturn.lastIndexOf(".map((territory"),
    beforeReturn.lastIndexOf(".map((district"),
    beforeReturn.lastIndexOf(".map((province"),
    beforeReturn.lastIndexOf(".map((state"),
    beforeReturn.lastIndexOf(".map((county"),
    beforeReturn.lastIndexOf(".map((municipality"),
    beforeReturn.lastIndexOf(".map((ward"),
    beforeReturn.lastIndexOf(".map((block"),
    beforeReturn.lastIndexOf(".map((lot"),
    beforeReturn.lastIndexOf(".map((plot"),
    beforeReturn.lastIndexOf(".map((parcel"),
    beforeReturn.lastIndexOf(".map((unit"),
    beforeReturn.lastIndexOf(".map((floor"),
    beforeReturn.lastIndexOf(".map((room"),
    beforeReturn.lastIndexOf(".map((desk"),
    beforeReturn.lastIndexOf(".map((seat"),
    beforeReturn.lastIndexOf(".map((position"),
    beforeReturn.lastIndexOf(".map((slot"),
    beforeReturn.lastIndexOf(".map((spot"),
    beforeReturn.lastIndexOf(".map((space"),
    beforeReturn.lastIndexOf(".map((place"),
    beforeReturn.lastIndexOf(".map((area"),
    beforeReturn.lastIndexOf(".map((zone"),
    beforeReturn.lastIndexOf(".map((region"),
    beforeReturn.lastIndexOf(".map((territory"),
    beforeReturn.lastIndexOf(".map((district"),
    beforeReturn.lastIndexOf(".map((province"),
    beforeReturn.lastIndexOf(".map((state"),
    beforeReturn.lastIndexOf(".map((county"),
    beforeReturn.lastIndexOf(".map((municipality"),
    beforeReturn.lastIndexOf(".map((ward"),
    beforeReturn.lastIndexOf(".map((block"),
    beforeReturn.lastIndexOf(".map((lot"),
    beforeReturn.lastIndexOf(".map((plot"),
    beforeReturn.lastIndexOf(".map((parcel"),
    beforeReturn.lastIndexOf(".map((unit"),
    beforeReturn.lastIndexOf(".map((floor"),
    beforeReturn.lastIndexOf(".map((room"),
    beforeReturn.lastIndexOf(".map((desk"),
    beforeReturn.lastIndexOf(".map((seat"),
    beforeReturn.lastIndexOf(".map((position"),
    beforeReturn.lastIndexOf(".map((slot"),
    beforeReturn.lastIndexOf(".map((spot"),
    beforeReturn.lastIndexOf(".map((space"),
  );
  
  return lastMapBeforeReturn > 0;
}

/**
 * Fix the "Page inside map" problem.
 * 
 * Strategy:
 * 1. Find the <Page ... > block
 * 2. Find the matching </Page>
 * 3. Extract the Page props
 * 4. Remove the <Page> and </Page> from their current location
 * 5. Find the component's top-level return (
 * 6. Wrap the entire JSX in <Page>
 */
function fixPagePlacement(content) {
  // Find <Page block
  const pageOpenMatch = content.match(/<Page\s[\s\S]*?>/);
  if (!pageOpenMatch) return null;
  
  const pageOpenStr = pageOpenMatch[0];
  const pageOpenIdx = content.indexOf(pageOpenStr);
  
  // Find the matching </Page>
  const pageCloseStr = "</Page>";
  // Find the </Page> that matches this <Page>
  // Walk forward from pageOpenIdx, counting <Page> and </Page>
  let depth = 0;
  let i = pageOpenIdx;
  let pageCloseIdx = -1;
  
  while (i < content.length) {
    if (content.startsWith("<Page", i) && (content[i + 5] === " " || content[i + 5] === "\n" || content[i + 5] === ">")) {
      depth++;
      i += 5;
    } else if (content.startsWith("</Page>", i)) {
      depth--;
      if (depth === 0) {
        pageCloseIdx = i;
        break;
      }
      i += 7;
    } else {
      i++;
    }
  }
  
  if (pageCloseIdx === -1) return null;
  
  // Extract the inner content between <Page> and </Page>
  const innerContent = content.substring(pageOpenIdx + pageOpenStr.length, pageCloseIdx);
  
  // Remove the <Page> and </Page> from their current location
  // Also remove any surrounding whitespace/newlines
  let before = content.substring(0, pageOpenIdx);
  let after = content.substring(pageCloseIdx + pageCloseStr.length);
  
  // Clean up the surrounding whitespace
  before = before.replace(/\s+$/, "\n");
  after = after.replace(/^\s+/, "\n");
  
  const withoutPage = before + innerContent + after;
  
  // Now find the component's top-level return (
  // Find the last 'return (' that is at the function body level (not inside a map)
  // Strategy: find the export default function, then find its return
  const exportMatch = withoutPage.match(/export\s+default\s+function\s+\w+[^{]*\{/);
  if (!exportMatch) return null;
  
  const funcBodyStart = withoutPage.indexOf(exportMatch[0]) + exportMatch[0].length;
  
  // Find the first 'return (' after funcBodyStart at the right indent level
  const returnRegex = /\n([ \t]*)return\s*\(/g;
  returnRegex.lastIndex = funcBodyStart;
  let m;
  let bestIdx = -1;
  let bestIndent = Infinity;
  while ((m = returnRegex.exec(withoutPage)) !== null) {
    const indentLen = m[1].length;
    if (indentLen <= bestIndent) {
      bestIndent = indentLen;
      bestIdx = m.index + 1; // position of 'return'
    }
  }
  
  if (bestIdx === -1) return null;
  
  // Find the opening paren of this return
  const returnSlice = withoutPage.substring(bestIdx);
  const parenOffset = returnSlice.indexOf("(");
  const jsxStart = bestIdx + parenOffset + 1;
  
  // Balance parens to find the closing paren
  let depth2 = 0;
  let j = bestIdx + parenOffset;
  let jsxEnd = -1;
  while (j < withoutPage.length) {
    if (withoutPage[j] === "(") depth2++;
    else if (withoutPage[j] === ")") {
      depth2--;
      if (depth2 === 0) {
        jsxEnd = j;
        break;
      }
    }
    j++;
  }
  
  if (jsxEnd === -1) return null;
  
  const innerJsx = withoutPage.substring(jsxStart, jsxEnd);
  
  // Wrap with <Page>
  const wrappedJsx = `\n        ${pageOpenStr}\n${innerJsx}\n        </Page>\n    `;
  
  return withoutPage.substring(0, jsxStart) + wrappedJsx + withoutPage.substring(jsxEnd);
}

// ─────────────────────────────────────────────────────────────────────────────
// Main
// ─────────────────────────────────────────────────────────────────────────────

const args = process.argv.slice(2);

let filesToProcess = [];

if (args[0] === "--all") {
  // Find all recently migrated files
  const appDir = join(ROOT, "app/(app)");
  const allPages = findPages(appDir);
  filesToProcess = allPages.filter(f => {
    const content = readFileSync(f, "utf-8");
    return content.includes("@/components/ui/Page") && hasPageInMapProblem(content);
  });
} else if (args[0]) {
  filesToProcess = [args[0]];
} else {
  console.error("Usage: node scripts/fix-page-placement.mjs <file-path> | --all");
  process.exit(1);
}

let fixed = 0;
let skipped = 0;
let errors = 0;

for (const filePath of filesToProcess) {
  try {
    const content = readFileSync(filePath, "utf-8");
    
    if (!hasPageInMapProblem(content)) {
      skipped++;
      continue;
    }
    
    const fixed_content = fixPagePlacement(content);
    if (!fixed_content) {
      console.log(`  ⏭️  ${filePath} — could not fix`);
      skipped++;
      continue;
    }
    
    writeFileSync(filePath, fixed_content, "utf-8");
    console.log(`  ✅ Fixed: ${filePath}`);
    fixed++;
  } catch (e) {
    console.log(`  ❌ Error: ${filePath}: ${e.message}`);
    errors++;
  }
}

console.log(`\nDone. Fixed: ${fixed} | Skipped: ${skipped} | Errors: ${errors}`);

File: scripts/fix-page-placement.py
#!/usr/bin/env python3
"""
Fix script: moves <Page> from inside .map() callbacks to the top-level return.

The migrate-misc-bulk.mjs script sometimes inserts <Page> inside a .map()
callback's return() instead of the component's top-level return(). This
script detects and fixes that pattern.

Usage:
  python3 scripts/fix-page-placement.py <file-path>
  python3 scripts/fix-page-placement.py --all
"""

import sys
import os
import re
import glob

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

def find_pages(directory):
    """Find all page.tsx files recursively."""
    result = []
    for root, dirs, files in os.walk(directory):
        for f in files:
            if f == "page.tsx":
                result.append(os.path.join(root, f))
    return result

def balance_parens(content, start):
    """Find the matching closing paren starting from 'start' (which should be '(')."""
    depth = 0
    i = start
    while i < len(content):
        if content[i] == '(':
            depth += 1
        elif content[i] == ')':
            depth -= 1
            if depth == 0:
                return i
        i += 1
    return -1

def balance_jsx_tags(content, start, tag_name):
    """Find the matching closing tag for an opening tag at 'start'."""
    open_tag = f"<{tag_name}"
    close_tag = f"</{tag_name}>"
    depth = 0
    i = start
    while i < len(content):
        if content[i:].startswith(open_tag) and (len(content) <= i + len(open_tag) or content[i + len(open_tag)] in ' \n\t>'):
            depth += 1
            i += len(open_tag)
        elif content[i:].startswith(close_tag):
            depth -= 1
            if depth == 0:
                return i
            i += len(close_tag)
        else:
            i += 1
    return -1

def extract_page_props(content):
    """Extract the <Page ...> opening tag."""
    m = re.search(r'<Page\s[\s\S]*?>', content)
    if not m:
        return None, -1, -1
    return m.group(0), m.start(), m.end()

def find_top_level_return(content):
    """Find the top-level return ( in the default export function."""
    # Find the export default function
    m = re.search(r'export\s+default\s+function\s+\w+[^{]*\{', content)
    if not m:
        return -1, -1
    
    func_body_start = m.end()
    
    # Find all 'return (' occurrences after func_body_start
    # The top-level one has the smallest indentation
    best_idx = -1
    best_indent = float('inf')
    
    for m2 in re.finditer(r'\n([ \t]*)return\s*\(', content[func_body_start:]):
        indent_len = len(m2.group(1))
        if indent_len <= best_indent:
            best_indent = indent_len
            best_idx = func_body_start + m2.start() + 1  # position of 'return'
    
    if best_idx == -1:
        return -1, -1
    
    # Find the opening paren
    paren_offset = content.index('(', best_idx)
    
    # Balance parens
    close_paren = balance_parens(content, paren_offset)
    
    return paren_offset + 1, close_paren  # jsxStart, jsxEnd

def has_page_in_map_problem(content):
    """Check if <Page> is inside a .map() callback."""
    if "@/components/ui/Page" not in content:
        return False
    
    page_idx = content.find("<Page")
    if page_idx == -1:
        return False
    
    before_page = content[:page_idx]
    
    # Find the last 'return (' before <Page
    last_return = before_page.rfind("return (")
    if last_return == -1:
        return False
    
    # Check if there's a .map( between the last 'return (' and <Page
    # that hasn't been closed
    before_return = before_page[:last_return]
    
    # Look for .map( patterns
    map_patterns = ['.map((', '.map(d =>', '.map((d,', '.map((kpi', '.map((item',
                    '.map((row', '.map((plan', '.map((step', '.map((q,', '.map((opt',
                    '.map((tab', '.map((s,', '.map((t,', '.map((c,', '.map((e,',
                    '.map((n,', '.map((f,', '.map((m,', '.map((p,', '.map((r,',
                    '.map((v,', '.map((i,', '.map((j,', '.map((k,', '.map((l,',
                    '.map((a,', '.map((b,', '.map((g,', '.map((h,', '.map((u,',
                    '.map((val', '.map((dept', '.map((emp', '.map((user',
                    '.map((ticket', '.map((award', '.map((survey', '.map((event',
                    '.map((benefit', '.map((card', '.map((link', '.map((metric',
                    '.map((stat', '.map((doc', '.map((file', '.map((role',
                    '.map((team', '.map((goal', '.map((task', '.map((action',
                    '.map((phase', '.map((stage', '.map((level', '.map((type',
                    '.map((cat', '.map((tag', '.map((badge', '.map((chip',
                    '.map((filter', '.map((col', '.map((node', '.map((edge',
                    '.map((point', '.map((bar', '.map((line', '.map((area',
                    '.map((pie', '.map((slice', '.map((segment', '.map((section',
                    '.map((group', '.map((cluster', '.map((bucket', '.map((bin',
                    '.map((range', '.map((period', '.map((month', '.map((week',
                    '.map((day', '.map((hour', '.map((year', '.map((quarter',
                    '.map((cycle', '.map((sprint', '.map((milestone',
                    '.map((location', '.map((place', '.map((site', '.map((venue',
                    '.map((city', '.map((country', '.map((region', '.map((zone',
                    '.map((territory', '.map((district', '.map((province',
                    '.map((state', '.map((county', '.map((unit', '.map((floor',
                    '.map((room', '.map((desk', '.map((seat', '.map((position',
                    '.map((slot', '.map((spot', '.map((space',
                    '.map((sw,', '.map((tkt,', '.map((kpi,', '.map((d,',
                    '.map((w,', '.map((x,', '.map((y,', '.map((z,',
                    '.map((o,', '.map((q,', '.map((idx,', '.map((index,',
                    '.map((num,', '.map((el,', '.map((elem,', '.map((entry,',
                    '.map((obj,', '.map((data,', '.map((info,', '.map((detail,',
                    '.map((record,', '.map((result,', '.map((output,',
                    '.map((input,', '.map((value,', '.map((key,', '.map((id,',
                    '.map((name,', '.map((label,', '.map((title,', '.map((text,',
                    '.map((desc,', '.map((msg,', '.map((note,', '.map((comment,',
                    '.map((item,', '.map((list,', '.map((arr,', '.map((set,',
                    '.map((map,', '.map((dict,', '.map((hash,', '.map((tree,',
                    '.map((graph,', '.map((path,', '.map((route,', '.map((url,',
                    '.map((link,', '.map((href,', '.map((src,', '.map((img,',
                    '.map((icon,', '.map((logo,', '.map((avatar,', '.map((photo,',
                    '.map((image,', '.map((video,', '.map((audio,', '.map((file,',
                    '.map((doc,', '.map((pdf,', '.map((csv,', '.map((json,',
                    '.map((xml,', '.map((html,', '.map((css,', '.map((js,',
                    '.map((ts,', '.map((tsx,', '.map((jsx,', '.map((md,',
                    '.map((txt,', '.map((log,', '.map((err,', '.map((warn,',
                    '.map((info,', '.map((debug,', '.map((trace,', '.map((fatal,',
                    '.map((critical,', '.map((error,', '.map((warning,',
                    '.map((success,', '.map((failure,', '.map((pending,',
                    '.map((active,', '.map((inactive,', '.map((enabled,',
                    '.map((disabled,', '.map((visible,', '.map((hidden,',
                    '.map((open,', '.map((closed,', '.map((expanded,',
                    '.map((collapsed,', '.map((selected,', '.map((checked,',
                    '.map((unchecked,', '.map((focused,', '.map((blurred,',
                    '.map((hovered,', '.map((pressed,', '.map((released,',
                    '.map((dragged,', '.map((dropped,', '.map((resized,',
                    '.map((moved,', '.map((scrolled,', '.map((zoomed,',
                    '.map((rotated,', '.map((flipped,', '.map((mirrored,',
                    '.map((inverted,', '.map((reversed,', '.map((sorted,',
                    '.map((filtered,', '.map((grouped,', '.map((aggregated,',
                    '.map((summarized,', '.map((totaled,', '.map((averaged,',
                    '.map((counted,', '.map((ranked,', '.map((scored,',
                    '.map((rated,', '.map((weighted,', '.map((normalized,',
                    '.map((scaled,', '.map((transformed,', '.map((converted,',
                    '.map((formatted,', '.map((parsed,', '.map((serialized,',
                    '.map((deserialized,', '.map((encoded,', '.map((decoded,',
                    '.map((encrypted,', '.map((decrypted,', '.map((hashed,',
                    '.map((signed,', '.map((verified,', '.map((validated,',
                    '.map((sanitized,', '.map((escaped,', '.map((unescaped,',
                    '.map((trimmed,', '.map((padded,', '.map((truncated,',
                    '.map((wrapped,', '.map((unwrapped,', '.map((flattened,',
                    '.map((nested,', '.map((merged,', '.map((split,',
                    '.map((joined,', '.map((concatenated,', '.map((appended,',
                    '.map((prepended,', '.map((inserted,', '.map((deleted,',
                    '.map((updated,', '.map((replaced,', '.map((swapped,',
                    '.map((moved,', '.map((copied,', '.map((cloned,',
                    '.map((duplicated,', '.map((removed,', '.map((added,',
                    '.map((created,', '.map((destroyed,', '.map((initialized,',
                    '.map((reset,', '.map((cleared,', '.map((emptied,',
                    '.map((filled,', '.map((loaded,', '.map((unloaded,',
                    '.map((mounted,', '.map((unmounted,', '.map((rendered,',
                    '.map((unrendered,', '.map((shown,', '.map((hidden,',
                    '.map((displayed,', '.map((hidden,', '.map((visible,',
                    '.map((invisible,', '.map((transparent,', '.map((opaque,',
                    '.map((blurred,', '.map((focused,', '.map((highlighted,',
                    '.map((dimmed,', '.map((brightened,', '.map((darkened,',
                    '.map((lightened,', '.map((colored,', '.map((decolored,',
                    '.map((styled,', '.map((unstyled,', '.map((themed,',
                    '.map((unthemed,', '.map((animated,', '.map((unanimated,',
                    '.map((transitioned,', '.map((untransitioned,',
                    '.map((transformed,', '.map((untransformed,',
                    '.map((translated,', '.map((untranslated,',
                    '.map((rotated,', '.map((unrotated,', '.map((scaled,',
                    '.map((unscaled,', '.map((skewed,', '.map((unskewed,',
                    '.map((flipped,', '.map((unflipped,', '.map((mirrored,',
                    '.map((unmirrored,', '.map((inverted,', '.map((uninverted,',
                    '.map((reversed,', '.map((unreversed,', '.map((sorted,',
                    '.map((unsorted,', '.map((filtered,', '.map((unfiltered,',
                    '.map((grouped,', '.map((ungrouped,', '.map((aggregated,',
                    '.map((unaggregated,', '.map((summarized,', '.map((unsummarized,',
                    '.map((totaled,', '.map((untotaled,', '.map((averaged,',
                    '.map((unaveraged,', '.map((counted,', '.map((uncounted,',
                    '.map((ranked,', '.map((unranked,', '.map((scored,',
                    '.map((unscored,', '.map((rated,', '.map((unrated,',
                    '.map((weighted,', '.map((unweighted,', '.map((normalized,',
                    '.map((unnormalized,', '.map((scaled,', '.map((unscaled,',
                    ]
    
    last_map = max((before_return.rfind(p) for p in map_patterns), default=-1)
    
    return last_map > 0

def fix_page_placement(content):
    """Fix the Page placement."""
    # Extract the <Page ...> opening tag
    page_open_str, page_open_idx, page_open_end = extract_page_props(content)
    if not page_open_str:
        return None
    
    # Find the matching </Page>
    page_close_idx = balance_jsx_tags(content, page_open_idx, "Page")
    if page_close_idx == -1:
        return None
    
    page_close_end = page_close_idx + len("</Page>")
    
    # Extract the inner content between <Page> and </Page>
    inner_content = content[page_open_end:page_close_idx]
    
    # Remove the <Page> and </Page> from their current location
    # We need to also remove surrounding whitespace
    before = content[:page_open_idx]
    after = content[page_close_end:]
    
    # Clean up surrounding whitespace
    before = before.rstrip() + "\n"
    after = "\n" + after.lstrip()
    
    without_page = before + inner_content + after
    
    # Find the top-level return
    jsx_start, jsx_end = find_top_level_return(without_page)
    if jsx_start == -1 or jsx_end == -1:
        return None
    
    inner_jsx = without_page[jsx_start:jsx_end]
    
    # Wrap with <Page>
    wrapped_jsx = f"\n        {page_open_str}\n{inner_jsx}\n        </Page>\n    "
    
    return without_page[:jsx_start] + wrapped_jsx + without_page[jsx_end:]

def main():
    args = sys.argv[1:]
    
    if not args:
        print("Usage: python3 scripts/fix-page-placement.py <file-path> | --all")
        sys.exit(1)
    
    if args[0] == "--all":
        app_dir = os.path.join(ROOT, "app/(app)")
        all_pages = find_pages(app_dir)
        files_to_process = []
        for f in all_pages:
            try:
                content = open(f).read()
                if has_page_in_map_problem(content):
                    files_to_process.append(f)
            except:
                pass
    else:
        files_to_process = [args[0]]
    
    fixed = 0
    skipped = 0
    errors = 0
    
    for file_path in files_to_process:
        try:
            content = open(file_path).read()
            
            if not has_page_in_map_problem(content):
                skipped += 1
                continue
            
            fixed_content = fix_page_placement(content)
            if not fixed_content:
                print(f"  ⏭️  {file_path} — could not fix")
                skipped += 1
                continue
            
            with open(file_path, 'w') as f:
                f.write(fixed_content)
            print(f"  ✅ Fixed: {file_path}")
            fixed += 1
        except Exception as e:
            print(f"  ❌ Error: {file_path}: {e}")
            errors += 1
    
    print(f"\nDone. Fixed: {fixed} | Skipped: {skipped} | Errors: {errors}")

if __name__ == "__main__":
    main()
